### PR TITLE
feat: suggests --all-sub-projects flag for multi-project gradle builds

### PR DESCRIPTION
**Pull request by kyegupov**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [x] Reviewed by Snyk internal team

#### What does this PR do?
Displays a suggestion to try `--all-sub-projects` flag.

#### What are the relevant tickets?
https://snyksec.atlassian.net/browse/BST-540

image

### CODE REVIEW

File: src/cli/commands/monitor.ts [hunk 1]
```diff
@@ -225,13 +231,20 @@ async function monitor(...args0: any[]): Promise<any> {
...
+  ) {
   const issues = res.licensesPolicy ? 'issues' : 'vulnerabilities';
   const humanReadableName = subProjectName ? `${res.path} (${subProjectName})` : res.path;
   let strOutput = chalk.bold.white('\nMonitoring ' + humanReadableName + '...\n\n') +
     (packageManager === 'yarn' ?
       'A yarn.lock file was detected - continuing as a Yarn project.\n' : '') +
       'Explore this snapshot at ' + res.uri + '\n\n' +
+    (advertiseSubprojectsCount ?
+      chalk.bold.white(`This project has multiple sub-projects (${advertiseSubprojectsCount}), ` +
+      'but, by default, you are only scanning one.\n' +
```
Comment (lili2311) [inline src/cli/commands/monitor.ts]: Can we drop the `but`,  perhaps:

`This project has multiple sub-projects(5), use --all-sub-projects flag to scan all sub projects.` this sounds more like our usual tone

File: src/cli/commands/monitor.ts [hunk 2]
```diff
@@ -136,15 +136,21 @@ async function monitor(...args0: any[]): Promise<any> {
       // a MultiDepRootsResult to an array of these.
 
       let perDepRootResults: SingleDepRootResult[] = [];
+      let advertiseSubprojectsCount: number | null = null;
       if (isMultiResult(inspectResult)) {
         perDepRootResults = inspectResult.depRoots.map(
           (depRoot) => ({plugin: inspectResult.plugin, package: depRoot.depTree}));
       } else {
+        if (!options['gradle-sub-project']
+          && inspectResult.plugin.meta
```
Comment (orsagie) [inline src/cli/commands/monitor.ts]:
```suggestion
          && _.get(inspectResult.plugin, 'meta.allDepRootNames', []).length > 1)
```

File: src/lib/snyk-test/run-test.ts [hunk 3]
```diff
@@ -176,15 +173,25 @@ async function getDepsFromPlugin(root, options): Promise<MultiRootsPackage> {
     if (!inspectRes.package.targetFile && inspectRes.plugin) {
       inspectRes.package.targetFile = inspectRes.plugin.targetFile;
     }
-    inspectRes.depRoots = [{depTree: inspectRes.package}];
+    // We are using "options" to store some information returned from plugin that we need to use later,
+    // but don't want to send to Registry in the Payload.
+    // TODO(kyegupov): decouple inspect and payload so that we don't need this hack
+    if (inspectRes.plugin.meta
```
Comment (orsagie) [inline src/lib/snyk-test/run-test.ts]:
```suggestion
    if (_.get(inspectResult.plugin, 'meta.allDepRootNames', []).length > 1)
```

File: src/cli/commands/monitor.ts (hunk 2 above)
Comment (kyegupov) [reply src/cli/commands/monitor.ts]: lodash is not type-checked and magical, I'd rather have verbose code here. If the type definitions change later (and they are likely to, see Language Initiative), any problems here will be easier to catch.

File: src/cli/commands/monitor.ts (hunk 1 above)
Comment (kyegupov) [reply src/cli/commands/monitor.ts]: Done.

**Comment by kyegupov** [comment]
Kicking the build

**Comment by kyegupov** [comment]
Kicking the build, again.

File: src/cli/commands/test.ts [hunk 4]
```diff
@@ -216,6 +217,15 @@ function displayResult(res, options) {
     dockerSuggestion += chalk.bold.white(docker.suggestionText);
   }
 
+  let multiProjAdvice = '';
+
+  if (options.advertiseSubprojectsCount) {
+    multiProjAdvice = chalk.bold.white(
+      `\n\nThis project has multiple sub-projects (${options.advertiseSubprojectsCount}), ` +
+      'but, by default, you are only scanning one.\n' +
```
Comment (lili2311) [inline src/cli/commands/test.ts]: also a but left over here, please update

Comment (kyegupov) [reply src/cli/commands/test.ts]: done

**Comment by kyegupov** [comment]
Kicking appveyor build :-/

**Pull request merged**